### PR TITLE
video(wayland): use both --icon and --icon-name for Zenity

### LINKED ISSUE #7895: SDL_ShowMessageBox (Wayland): Zenity --icon-name is deprecated
On my system (Fedora 38 with Zenity 3.92.0), `--icon-name` is deprecated in favor of `--icon`. This is completely broken until 3.99.0 where the option is added back with a deprecation notice.

### PR DESCRIPTION
**Pull request by takase1121**
This PR tries to use --icon for Zenity and fall back to --icon-name when it fails.

## Description

Many distros ship an older version of Zenity that supports GTK3, while some distros ship newer version of Zenity which uses libadwaita. These different version has different flags and only the latest version (3.99.0) provides a safe fallback for `--icon-name`.

## Existing Issue(s)

Closes #7895.

### CODE REVIEW

**Comment by slouken** [comment]
What testing have you done of this change?
Does zenity show an error message the first time if it doesn’t support the command line option?

**Comment by Guldoman** [comment]
#7781 has some more info, and is replaced by this PR.

**Comment by takase1121** [comment]
> What testing have you done of this change?

I compiled an older version of Zenity that doesn't support the option and add it to my path. After that, I run `testmessage` to verify if it is working.

> Does zenity show an error message the first time if it doesn’t support the command line option?

Yes it does. If you want to avoid this, I can attempt to parse the version number returned by zenity before invoking it with the correct command, but I think this is quite overkill and may make it complex.

File: src/video/wayland/SDL_waylandmessagebox.c [hunk 1]
```diff
@@ -188,4 +191,16 @@ int Wayland_ShowMessageBox(const SDL_MessageBoxData *messageboxdata, int *button
     }
 }
 
+int Wayland_ShowMessageBox(const SDL_MessageBoxData *messageboxdata, int *buttonid)
+{
+    int i, err;
+
+    for (i = 0; i < sizeof(icon_name_options) / sizeof(*icon_name_options); ++i) {
```
Comment (Susko3) [inline src/video/wayland/SDL_waylandmessagebox.c:198]: Should probably use `SDL_arraysize` macro.

**Comment by slouken** [comment]
@Kontrabant, what's the use case for Zenity here, should we be using dbus portal API instead as a more portable dialog solution?

**Comment by Kontrabant** [comment]
There is no dbus portal for generic message boxes. The two options for native-looking dialogs are basically Zenity, or dynamically loading GTK or Qt and using that to create a message box.

**Comment by Guldoman** [comment]
If this fix won't be backported to SDL2, it might be a good idea to wait until we're closer to a proper SDL3 release, in case no distro will be shipping the broken version at that time.

**Comment by slouken** [comment]
It probably should be backported to SDL2, it sounds like there are currently distros where wayland messageboxes don't work at all right now.

**Comment by Guldoman** [comment]
Yeah, fair.
I wonder if it would be worth it to propose a dialogs portal.

**Comment by takase1121** [comment]
So, going forward do I need to parse Zenity versions, or is the current approach of running it twice okay?

**Comment by Kontrabant** [comment]
> Yeah, fair. I wonder if it would be worth it to propose a dialogs portal.

No harm in asking, but I suspect that it would be shot down. Portals tend to exist to provide access control to system resources for sandboxed applications as opposed to being a generic GUI toolkit.  Unless there is a security case to be made for typical Yes/No/OK modal dialogs, I doubt this would be something that the portal or desktop maintainers are interested in.

**Comment by slouken** [comment]
> So, going forward do I need to parse Zenity versions, or is the current approach of running it twice okay?

Yes, let's go ahead and parse zenity versions. It's ugly, but it's an ugly problem.

Thanks!

**Comment by Guldoman** [comment]
Wouldn't just retrying be better? Best case Zenity runs only once, worst case it runs twice.
Also considering that the vast majority of Zenity versions (and newer versions in general) will fall down the good case path.

**Comment by slouken** [comment]
Showing an error message on the console every time a messagebox pops up isn't great either. I'm inclined to have a separate function that once per run will detect the zenity version and set up the appropriate option.

I could be convinced though, thoughts everyone?

**Comment by takase1121** [comment]
Alright; I've implemented version parsing and now the code calls Zenity with the right argument. I didn't cache the version number because I'm unsure how one would do that in SDL (static variables?).

File: src/video/wayland/SDL_waylandmessagebox.c [hunk 2]
```diff
@@ -32,175 +32,218 @@
...
+
+    if (pipe(fd_pipe) != 0) { /* create a pipe */
+        return SDL_SetError("pipe() failed: %s", strerror(errno));
+    }
+
+    /* https://gitlab.gnome.org/GNOME/zenity/-/commit/c686bdb1b45e95acf010efd9ca0c75527fbb4dea
+     * This commit removed --icon-name without adding a deprecation notice.
+     * We need to handle it gracefully, otherwise no message box will be shown.
+     */
+    argv[argc++] = zenity_major >= 3 && zenity_minor >= 90 ? "--icon" : "--icon-name";
```
[the remark this answers is not in the document]
Comment (takase1121) [reply src/video/wayland/SDL_waylandmessagebox.c:148]: You're right!

**Comment by slouken** [comment]
Merged, thanks!

**Comment by slouken** [comment]
And this is now in for SDL 2.28.2

**Pull request merged**